plugins/org.eclipse.n4js/src/org/eclipse/n4js/typesystem/ExpectedTypeJudgment.java

```diff
@@ -287,6 +290,29 @@ public TypeRef casePostfixExpression(PostfixExpression e) {
 
 		@Override
 		public TypeRef caseUnaryExpression(UnaryExpression e) {
+			Expression innerExpression = e.getExpression();
+			TypeRef typeRef = ts.tau(innerExpression);
+
+			if (innerExpression instanceof UnaryExpression) {
+				typeRef = ts.tau(((UnaryExpression) innerExpression).getExpression());
+			}
+
+			boolean isNumber = RuleEnvironmentExtensions.isNumeric(G, typeRef);
```

Comment:
tau is not that cheap! Could you refactor the code so that tau (and typeRef etc.) are only used if necessary. E.g., first check the operator (I would assume INV) and if it is INV only then examine the rest  -- also taking care of the order in which things are checked and computed.